src/canister/individual_user_template/src/api/cdao/delete_all_sns_creator_token.rs

```diff
@@ -54,6 +57,72 @@ pub fn delete_sns_creator_token(deployed_canisters: &DeployedCdaoCanisters) {
             )
             .await;
         }
-        uninstall_code_and_return_empty_canisters_to_subnet_backup_pool(canister_ids).await;
+
+        let needs_to_be_deleted =
+            anonymous_session || is_sns_canister_in_invalid_state(deployed_canisters.swap).await;
+
+        if needs_to_be_deleted {
+            uninstall_code_and_return_empty_canisters_to_subnet_backup_pool(canister_ids).await;
+            CANISTER_DATA.with_borrow_mut(|canister_data| {
+                canister_data
+                    .cdao_canisters
+                    .retain(|key| *key != deployed_canisters);
+            });
+
+            ic_cdk::println!(
+                "Deleting creator SNS canisters with Root Canister Id {}",
+                deployed_canisters.root.to_text()
+            )
+        } else {
+            update_controllers_for_ledger_and_swap(deployed_canisters).await
+        }
     });
 }
+
+pub async fn is_sns_canister_in_invalid_state(swap_canister_id: Principal) -> bool {
+    let swap_canister = swap::Service(swap_canister_id);
+
+    let swap_lifecycle_res = swap_canister.get_lifecycle(GetLifecycleArg {}).await;
+
+    match swap_lifecycle_res {
+        Ok((lifecycle,)) => {
+            if let Some(lifecycle) = lifecycle.lifecycle {
+                lifecycle == 4
+            } else {
+                false
+            }
+        }
+        Err(_e) => false,
+    }
+}
+
+pub async fn update_controllers_for_ledger_and_swap(deployed_canister: DeployedCdaoCanisters) {
+    let global_super_admin_principal = Principal::from_text(GLOBAL_SUPER_ADMIN_USER_ID).unwrap();
+
+    let _ = update_settings(UpdateSettingsArgument {
+        canister_id: deployed_canister.ledger,
+        settings: CanisterSettings {
+            controllers: Some(vec![
+                global_super_admin_principal,
+                id(),
+                deployed_canister.root,
+            ]),
+            ..Default::default()
+        },
+    })
+    .await;
+
+    let _ = update_settings(UpdateSettingsArgument {
+        canister_id: deployed_canister.swap,
+        settings: CanisterSettings {
+            controllers: Some(vec![
+                global_super_admin_principal,
+                id(),
+                deployed_canister.root,
+                ic_nns_constants::ROOT_CANISTER_ID.into(),
+            ]),
+            ..Default::default()
+        },
+    })
+    .await;
+}
```

Comment:
reason we are calling update_settings 2 times ? can't we add all controllers at once ?

Reply:
nvm. my bad